### DISCUSSION TITLE
How to monitor codes sent by tablet?

### DISCUSSION THREAD

**Discussion by kzembower** [xournalpp/xournalpp#4234] [General]
Hello, all,

I'm finally ready to start implementing a plugin for my Huion 1060PLUS (8192) drawing tablet, as I asked back in https://github.com/xournalpp/xournalpp/discussions/3296#discussion-3492286 (Yes, I know I asked about this a while ago. I'm an old man, and slow.)

I'm working with this tablet on a Ubuntu 22.04 jammy workstation. I need some help getting started. One of my first questions is exactly what is the tablet sending when one of the define-able buttons is pressed.

Would anyone point me to a tool to use to read the codes generated by button presses on the tablet? I've learned about usbmon and USBmon and usbdump, but want to ask here if this is the way to go, as they all seem pretty low-level and have a steep learning curve.

I guess I could hook the tablet to a Windows 11 system I have, and use the Huion tools. The sent codes should remain the same from system to system, I assume. Would this be the easiest way to go?

Thanks for your help and mentoring. I hope your efforts will help me create a new plugin for using tablets of this type.

-Kevin

**Comment by rolandlo** [comment]
The idea was to use `xsetwacom` to map the buttons of your tablet to key strokes (e.g. `CTRL+SHIFT+ALT+a`, `CTRL+SHIFT+ALT+b` etc.) like in answer 6 and 7 of this [Krita issue](https://krita-artists.org/t/hotkeys-on-huion-tablet/3222/6), so that each time you press the button, the corresponding keystroke is emitted.
Plugin callback functions get activated through keyboard accelerators as defined in the `initUi` function, see [guide](https://xournalpp.github.io/guide/plugins/plugins/#plugin-structure). 
On Ubuntu 22.04 you should use the Xorg session (at the login screen), not Wayland. Otherwise you have to find an alternative for `xsetwacom` on Wayland.
It could also be that your tablet driver comes with a GUI to assign these keystrokes to the tablet buttons. That would simplify things, since you can avoid using a command line tool like `xsetwacom` in that case.

**Reply by kzembower** [reply]
@rolandlo, thanks very much for your help.

Can you say more about "...you should use the Xorg session (at the login screen), not Wayland." I got this output when I ran `xsetwacom --list`:
```
$ xsetwacom --list
Wayland devices found but this tool is incompatible with Wayland. See
https://github.com/linuxwacom/xf86-input-wacom/wiki/Wayland
$
```
Got this output, too:
```
$ echo $XDG_SESSION_TYPE
wayland
$
```
I've just been exploring the differences between Xorg and Wayland, but don't know much yet. What do you mean by "use Xorg session at the login screen"? How is this done?

Thanks, again, for your help and guidance.

-Kevin

**Reply by kzembower** [reply]
I was able to switch to the X11 session, but now the tablet doesn't show up in `xsetwacom --list`, even though it shows up in `lsusb` at the bus 003 7, 8 and 9 devices, and even though the tablet works as a mouse in a terminal:

```
$ echo $XDG_SESSION_TYPE
x11
$ xsetwacom --list
$ lsusb
Bus 002 Device 002: ID 8087:8000 Intel Corp. Integrated Rate Matching Hub
Bus 002 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 001 Device 002: ID 8087:8008 Intel Corp. Integrated Rate Matching Hub
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 004 Device 002: ID 0bc2:2343 Seagate RSS LLC Portable
Bus 004 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 003 Device 002: ID 05e3:0604 Genesys Logic, Inc. USB 1.1 Hub
Bus 003 Device 006: ID 25a7:fa10 Areson Technology Corp 2.4G Wireless Receiver
Bus 003 Device 009: ID 058f:6366 Alcor Micro Corp. Multi Flash Reader
Bus 003 Device 008: ID 256c:006e  
Bus 003 Device 007: ID 058f:6254 Alcor Micro Corp. USB Hub
Bus 003 Device 005: ID 04fe:0006 PFU, Ltd Happy Hacking Keyboard Lite2
Bus 003 Device 004: ID 04fe:0008 PFU, Ltd Generic USB Hub
Bus 003 Device 003: ID 1bcf:2284 Sunplus Innovation Technology Inc. Full HD webcam
Bus 003 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
$ 
```
Any thoughts? Will `xsetwacom` commands still work if `xsetwacom --list` doesn't find the tablet?

-Kevin

**Comment by bhennion** [comment]
> Would anyone point me to a tool to use to read the codes generated by button presses on the tablet? I've learned about usbmon and USBmon and usbdump, but want to ask here if this is the way to go, as they all seem pretty low-level and have a steep learning curve.

A very useful tool for input devices monitoring is `evtest`. It simply dumps on the screen every event from a selected input device.

**Reply by kzembower** [reply]
@bhennion, thanks very much for your help. I didn't know of 'evtest' at all, and probably wouldn't have learned about it without your suggestion. I'm not sure I'll need it now, using the procedures from @rolandlo above, but if I need it, I think the evtest will do just what I need.

Thanks, again, for your help.

-Kevin
